lib/srv/db/cloud/users/user.go

```diff
@@ -0,0 +1,215 @@
+func (u *baseUser) GetID() string {
+	return u.secretKey
+}
+
+// GetInDatabaseName returns in-database username for the user.
+func (u *baseUser) GetInDatabaseName() string {
+	return u.inDatabaseName
+}
+
+// Setup preforms any setup necessary like creating password secret.
+func (u *baseUser) Setup(ctx context.Context) error {
+	u.log.Debugf("Setting up user %v", u)
+
+	newPassword, err := genRandomPassword(u.maxPasswordLength)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = u.secrets.Create(ctx, u.secretKey, newPassword)
+	if err != nil {
+		if trace.IsAlreadyExists(err) {
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+
+	if u.modifyUserFunc != nil {
+		return u.modifyUserFunc(ctx, "", newPassword)
+	}
+	return nil
+}
+
+// Teardown performs any teardown necessary like deleting password secret.
+func (u *baseUser) Teardown(ctx context.Context) error {
+	u.log.Debugf("Tearing down user %v", u)
+
+	err := trace.Wrap(u.secrets.Delete(ctx, u.secretKey))
+	if err != nil {
+		// The secret may have been removed by another agent already.
+		if trace.IsNotFound(err) {
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// GetPassword returns the password used for database login.
+func (u *baseUser) GetPassword(ctx context.Context) (string, error) {
+	// Use current/latest version for login.
+	if !u.usePreviousPasswordForLogin {
+		return u.getPassword(ctx, secrets.CurrentVersion)
+	}
+
+	// User previous version for login.
+	password, err := u.getPassword(ctx, secrets.PreviousVersion)
+	if err != nil {
+		// Rare case check when there is only one version at the moment. Do a
+		// second get to use the current version.
+		//
+		// It is also possible someone else has deleted the secret completely.
+		// In that case the next rotate password will handle it by recreating
+		// the secret.
+		if trace.IsNotFound(err) {
+			return u.getPassword(ctx, secrets.CurrentVersion)
+		}
+		return "", trace.Wrap(err)
+	}
+	return password, nil
+}
+
+// getPassword returns the password used for database login.
+func (u *baseUser) getPassword(ctx context.Context, version string) (string, error) {
+	value, err := u.secrets.GetValue(ctx, u.secretKey, version)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return value.Value, nil
+}
+
+// RotatePassword rotates user's password.
+func (u *baseUser) RotatePassword(ctx context.Context) error {
+	currentValue, err := u.secrets.GetValue(ctx, u.secretKey, secrets.CurrentVersion)
+	if err != nil {
+		// Rare case check when someone else has deleted the secret.
+		if trace.IsNotFound(err) {
+			return u.Setup(ctx)
+		}
+
+		return trace.Wrap(err)
+	}
+
+	// The password is up-to-date. Nothing to do.
+	expiresAt := currentValue.CreatedAt.Add(u.secretTTL)
+	if u.clock.Now().Before(expiresAt) {
+		return nil
+	}
+
+	u.log.Debugf("Updating password for user %v", u)
+	newPassword, err := genRandomPassword(u.maxPasswordLength)
+	if err != nil {
```

Comment:
This will mostly work fine but there is a some cornet case when n agents can run u.secrets.GetValue( simultaneously (for example after cluster downtime) and obtain not to up to day password and run N times `u.secrets.PutValue` and drain the secret rotation limit. This is probably rare corner case though could you left a comment describing this issue ?

Reply:
Right, Thanks explantation. I have missed `ClientRequestToken` behaviour. In that case my comment seems to be not relevant.